pkg/utils/refurbisher.go

```diff
@@ -0,0 +1,31 @@
+package utils
+
+import (
+	"context"
+	"sync"
+)
+
+// Refurbisher handles serial tasks obsoleting each other.
+type Refurbisher struct {
+	mtx    sync.Mutex
+	cancel func()
+}
+
+// NewRefurbisher returns a new Refurbisher.
+func NewRefurbisher() *Refurbisher {
+	return &Refurbisher{cancel: func() {}}
+}
+
+// Do runs f after cancelling the previous f (if still running).
+func (r *Refurbisher) Do(ctx context.Context, f func(context.Context) error) error {
+	var childCtx context.Context
+
+	r.mtx.Lock()
+	cancel := r.cancel
+	childCtx, r.cancel = context.WithCancel(ctx)
+	r.mtx.Unlock()
+
+	cancel()
+
+	return f(childCtx)
```

Comment:
I think `go f()` here would make more sense than a `go Do()` by the caller.

This would prevent operations started after each other to be reordered, so
```go
go r.Do(ctx, f)
go r.Do(ctx, g)
```
could actually run `f` last.

Additionally, I think it would make sense to ensure that only one function is running at once, otherwise, the context for `f` might be cancelled, but it could still complete an action that might conflict with `g`.

Reply:
I don't think there's a guarantee on how fast/in which order goroutines are starting. So basically you're betting that `go r.Do(ctx, f)` locks the mutex first. If you're performing the operations protected by the mutex before spawning the goroutines, there's a guaranteed order (if both `r.Do()` calls are performed from the same goroutine or are synchronized by other means).

Reply:
At the moment probably none, all seem to be doing some (indirect) network I/O, so that should be enough time for another goroutine to be scheduled and acquire a mutex. It's just one these "change a caller and you might get one a one in a million executions bugs that nobody can figure out".

But after looking at all the callers once again: changing this would mean means you can no longer combine this with an errgroup as easily. We've already got `ErrgroupReceive`, so combining this with returning a channel and this will also work there.